File: ios/Cove/HomeScreens/ListWalletsScreen.swift
//
//  ListWalletsScreen.swift
//  Cove
//
//  Created by Praveen Perera on 6/30/24.
//

import SwiftUI

struct ListWalletsScreen: View {
    @Environment(AppManager.self) private var app
    @State var wallets: [WalletMetadata]
    @Environment(\.navigate) private var navigate

    init() {
        do {
            wallets = try Database().wallets().all()
            Log.debug("Wallets: \(wallets)")
        } catch {
            Log.error("Failed to get wallets \(error)")
            wallets = []
        }
    }

    var body: some View {
        FullPageLoadingView()
            .onAppear {
                do {
                    if let wallet = wallets.first {
                        return try app.rust.selectWallet(id: wallet.id)
                    }
                } catch {
                    app.loadAndReset(to: Route.newWallet(.select))
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

#Preview {
    ListWalletsScreen()
        .environment(AppManager.shared)
}

File: android/app/src/main/java/org/bitcoinppl/cove/ListWalletsScreen.kt
package org.bitcoinppl.cove

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Modifier
import org.bitcoinppl.cove.components.FullPageLoadingView
import org.bitcoinppl.cove_core.*
import org.bitcoinppl.cove_core.types.*

/**
 * list wallets screen - auto-selects first wallet or navigates to add wallet
 * ported from iOS ListWalletsScreen.swift
 */
@Composable
fun ListWalletsScreen(
    app: AppManager,
    modifier: Modifier = Modifier,
) {
    // show loading indicator
    FullPageLoadingView(modifier = modifier)

    // on appear, check for wallets and navigate
    LaunchedEffect(Unit) {
        try {
            val wallets = Database().wallets().all()
            android.util.Log.d("ListWalletsScreen", "Found ${wallets.size} wallets")

            val firstWallet = wallets.firstOrNull()
            if (firstWallet != null) {
                // select the first wallet
                app.rust.selectWallet(firstWallet.id)
            } else {
                // no wallets, go to add wallet screen
                app.loadAndReset(RouteFactory().newWalletSelect())
            }
        } catch (e: Exception) {
            android.util.Log.e("ListWalletsScreen", "Failed to get wallets", e)
            // on error, navigate to add wallet
            app.loadAndReset(RouteFactory().newWalletSelect())
        }
    }
}
